### PR TITLE
Use template class for in-memory data.

### PR DESCRIPTION
**Pull request by yxue**
Fixes https://github.com/open-telemetry/opentelemetry-cpp/issues/1493

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h [hunk 1]
```diff
@@ -19,14 +21,16 @@ const size_t MAX_BUFFER_SIZE = 100;
  * A in memory exporter that switches a flag once a valid recordable was received
  * and keeps track of all received spans in memory.
  */
+using InMemorySpanData = exporter::memory::InMemoryData<sdk::trace::SpanData>;
```
Comment (lalitb) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:24]: should we instead have `InMemorySpanData` as a class to inherit from `InMemoryData`? And also have `InMemorySpanData::GetSpans()` method. This way applications using span memory exporter won't need to change.

Comment (owent) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:24]: Agree

Comment (yxue) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h]: Thanks! Updated the PR to inherit from the InMemoryData. PTAL.

**Review by lalitb — APPROVED** [review]
LGTM

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter_factory.h [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 #pragma once
 
-#include "opentelemetry/exporters/memory/in_memory_span_data.h"
+#include "opentelemetry/exporters/memory/in_memory_span_exporter.h"
```
Comment (marcalff) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter_factory.h:6]: Why was this include changed ?

The whole point of the factory class is to hide the implementation,
so there should be no dependencies on the in_memory_span_exporter.h header.

**Review by marcalff — CHANGES_REQUESTED** [review]

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter_factory.h (hunk 2 above)
Comment (lalitb) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter_factory.h:6]: Good catch. I think `InMemorySpanData` class definition should be moved to `in_memory_span_data.h`.

Comment (yxue) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter_factory.h]: Thanks! Comments resolved.

**Review by owent — APPROVED** [review]
LGTM

**Review by marcalff — APPROVED** [review]
LGTM, thanks for the fixes.

**Pull request merged**